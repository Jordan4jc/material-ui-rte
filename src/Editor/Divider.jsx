import React from "react"
import Divider from "@material-ui/core/Divider"

const DividerComponent = props => {
  return <Divider />
}

export default DividerComponent
