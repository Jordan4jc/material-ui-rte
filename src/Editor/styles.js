const styles = theme => ({
  toolbar: {
    marginBottom: theme.spacing(2),
    display: "flex",
    justifyContent: "space-between"
  },
  editor: {
    padding: theme.spacing(2)
  }
})

export default styles
