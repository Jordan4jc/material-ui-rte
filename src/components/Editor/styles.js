const styles = theme => ({
  paper: {
    paddingBottom: theme.spacing(2)
  },
  toolbar: {
    display: "flex",
    justifyContent: "space-between",
    backgroundColor: theme.palette.grey[200]
  },
  editor: {
    margin: theme.spacing(2),
    marginBottom: 0,
    padding: theme.spacing(2),
    backgroundColor: theme.palette.grey[100],
    borderRadius: theme.shape.borderRadius,
    cursor: "text"
  }
})

export default styles
