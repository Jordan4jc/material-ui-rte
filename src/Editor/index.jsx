import React, { Component, Fragment } from "react"
import PropTypes from "prop-types"
import {Editor, EditorState, RichUtils} from "draft-js"
import { withStyles } from "@material-ui/styles"
import Paper from '@material-ui/core/Paper';
import BlockStyleControl from "../BlockStyleControl";
import styles from "./styles"
import InlineStyleControls from "../InlineStyleControls";

export class EditorComponent extends Component {
  static propTypes = {

  }

  state = {
    blockStyle: {},
    editorState: EditorState.createEmpty()
  }

  onEditorChange = editorState => this.setState({ editorState })

  handleBlockStyleChange = style => {
    this.setState({
      blockStyle: style
    })
    this.onEditorChange(
      RichUtils.toggleBlockType(
        this.state.editorState,
        style
      )
    );
  }

  render() {
    const { classes } = this.props
    return (
      <Fragment>
        <section className={classes.toolbar}>
          <BlockStyleControl onChange={this.handleBlockStyleChange} />
          <InlineStyleControls onChange={this.handleInlineStyleChange} />
        </section>
        <Paper className={classes.editor}>
          <Editor editorState={this.state.editorState} onChange={this.onEditorChange} />
        </Paper>
      </Fragment>
    )
  }
}

export default withStyles(styles, { withTheme: true })(EditorComponent)
