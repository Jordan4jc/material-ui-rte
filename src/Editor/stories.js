import React from "react"
import { createMuiTheme } from "@material-ui/core/styles"
import { ThemeProvider } from "@material-ui/styles"
// Import the storybook libraries
import { storiesOf } from "@storybook/react"
import { action } from "@storybook/addon-actions"
// Import our component from this folder
import Editor from "./index"

const theme = createMuiTheme({
  palette: {
    text: {
      primary: "#f00"
    }
  }
})

storiesOf("Editor")
  .add("Basic", () => <Editor onChange={action("editor content changed")} />)
  .add("Customized Controls", () => (
    <Editor
      onChange={action("editor content changed")}
      blockStyleControls={[
        "paragraph",
        "header-one",
        "header-two",
        "header-three"
      ]}
      imageUploadControl={false}
      alignmentControls={["left", "center"]}
    />
  ))
