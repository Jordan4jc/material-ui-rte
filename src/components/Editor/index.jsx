import React, { useState, useRef, useEffect } from "react"
import PropTypes from "prop-types"
import Draft, {
  EditorState,
  convertFromRaw,
  convertToRaw,
  AtomicBlockUtils,
  getDefaultKeyBinding,
  KeyBindingUtil
} from "draft-js"
import RichUtils from "../RichUtils"
import blockStyleFn from "../blockStyleFn"
import blockRenderMap from "../blockRenderMap"
import Editor from "draft-js-plugins-editor"
import createBlockBreakoutPlugin from "draft-js-block-breakout-plugin"
import createImagePlugin from "draft-js-image-plugin"
import createDividerPlugin from "draft-js-divider-plugin"
import { makeStyles } from "@material-ui/styles"
import { StylesProvider, createGenerateClassName } from "@material-ui/styles"
import Paper from "@material-ui/core/Paper"
import Toolbar from "@material-ui/core/Toolbar"
import BlockStyleControl from "../BlockStyleControl"
import styles from "./styles"
import InlineStyleControls from "../InlineStyleControls"
import AlignmentControls from "../AlignmentControls"
import ListControls from "../ListControls"
import DividerControl from "../DividerControl"
import Divider from "./Divider"
import ImageUploadControl from "../ImageUploadControl"

// SETUP KEYBINDINGS
const { hasCommandModifier } = KeyBindingUtil

const generateClassName = createGenerateClassName({
  productionPrefix: "mur"
})

const useStyles = makeStyles(styles, {
  name: "MuiRte-Editor"
})

const extendedBlockRenderMap = Draft.DefaultDraftBlockRenderMap.merge(
  blockRenderMap
)

const blockBreakoutPlugin = createBlockBreakoutPlugin()
const dividerPlugin = createDividerPlugin({
  blockType: "divider",
  component: Divider
})

const imagePlugin = createImagePlugin()

const { addDivider } = dividerPlugin

const plugins = [blockBreakoutPlugin, dividerPlugin, imagePlugin]

// SETUP TYPES FOR PROP CHECKING
const availableBlockStyles = [
  "paragraph",
  "blockquote",
  "header-one",
  "header-two",
  "header-three",
  "header-four",
  "header-five",
  "header-six"
]
const availableInlineStyles = ["bold", "italic", "underline"]
const availableListTypes = ["ordered-list", "unordered-list"]
const availableAlignmentStyles = ["left", "center", "right"]

const myKeyBindingFn = e => {
  // hook up alignment shortcuts
  if (e.keyCode === 76 /* `L` key */ && hasCommandModifier(e)) {
    return "align-left"
  }
  if (e.keyCode === 69 /* `E` key */ && hasCommandModifier(e)) {
    return "align-center"
  }
  if (e.keyCode === 82 /* `R` key */ && hasCommandModifier(e)) {
    return "align-right"
  }
  return getDefaultKeyBinding(e)
}

export const EditorComponent = props => {
  const [editorState, setEditorState] = useState(EditorState.createEmpty())
  useEffect(() => {
    if (props.value) {
      setEditorState(EditorState.createWithContent(convertFromRaw(props.value)))
    }
  }, [props.value])
  const editorRef = useRef(null)

  const classes = useStyles()

  const onEditorChange = editorState => {
    setEditorState(editorState)
    if (props.hasOwnProperty("onChange")) {
      props.onChange(convertToRaw(editorState.getCurrentContent()))
    }
  }

  const handleBlockStyleChange = style => {
    onEditorChange(RichUtils.toggleBlockType(editorState, style))
  }

  const handleAlignmentStyleChange = style => {
    onEditorChange(RichUtils.toggleAlignment(editorState, style))
  }

  const handleInlineStyleChange = style => {
    onEditorChange(RichUtils.toggleInlineStyle(editorState, style))
  }

  const handleDividerControlClick = () => {
    onEditorChange(addDivider(editorState))
  }
  const handleFile = file => {
    console.log(file)
    const contentState = editorState.getCurrentContent()
    const contentStateWithEntity = contentState.createEntity(
      "IMAGE",
      "IMMUTABLE",
      { src: file.preview, file }
    )
    const entityKey = contentStateWithEntity.getLastCreatedEntityKey()
    const newEditorState = EditorState.set(editorState, {
      currentContent: contentStateWithEntity
    })
    const editorStateWithImage = AtomicBlockUtils.insertAtomicBlock(
      newEditorState,
      entityKey,
      " "
    )
    onEditorChange(editorStateWithImage)
  }

  const handlekeyCommand = command => {
    if (command.indexOf("align-") > -1 && props.alignmentControls !== false) {
      let allowedControls = []
      if (props.alignmentControls === true) {
        allowedControls = availableAlignmentStyles
      } else {
        allowedControls = props.alignmentControls
      }
      const alignment = command.replace("align-", "")
      if (allowedControls.indexOf(alignment) > -1) {
        handleAlignmentStyleChange(alignment.toUpperCase())
        return "handled"
      }
      return "not-handled"
    }
    return "not-handled"
  }

  return (
    <StylesProvider generateClassName={generateClassName}>
      <Paper className={classes.paper}>
        <Toolbar className={classes.toolbar}>
          {props.blockStyleControls !== false && (
            <BlockStyleControl
              controls={props.blockStyleControls}
              onChange={handleBlockStyleChange}
            />
          )}
          {props.dividerControl !== false && (
            <DividerControl onClick={handleDividerControlClick} />
          )}
          {props.imageUploadControl !== false && (
            <ImageUploadControl onFile={handleFile} />
          )}
          {props.inlineStyleControls !== false && (
            <InlineStyleControls
              controls={props.inlineStyleControls}
              editorState={editorState}
              onChange={handleInlineStyleChange}
            />
          )}
          {props.listControls !== false && (
            <ListControls
              controls={props.listControls}
              editorState={editorState}
              onChange={handleBlockStyleChange}
            />
          )}
          {props.alignmentControls !== false && (
            <AlignmentControls
              controls={props.alignmentControls}
              editorState={editorState}
              onChange={handleAlignmentStyleChange}
            />
          )}
        </Toolbar>
        <div
          className={classes.editor}
          onClick={() => editorRef.current.focus()}
        >
          <Editor
            ref={editorRef}
            handleKeyCommand={handlekeyCommand}
            keyBindingFn={myKeyBindingFn}
            blockStyleFn={blockStyleFn}
            blockRenderMap={extendedBlockRenderMap}
            editorState={editorState}
            onChange={onEditorChange}
            plugins={plugins}
          />
        </div>
      </Paper>
    </StylesProvider>
  )
}

EditorComponent.propTypes = {
  blockStyleControls: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.arrayOf(function(
      propValue,
      key,
      componentName,
      location,
      propFullName
    ) {
      return propValue.some(p => {
        if (availableBlockStyles.indexOf(p) === -1) {
          return new Error(
            `Invalid prop ${propFullName} supplied to ${componentName}. Should be one of ${availableBlockStyles.join(
              " | "
            )}`
          )
        } else {
          return true
        }
      })
    })
  ]),
  dividerControl: PropTypes.bool,
  imageUploadControl: PropTypes.bool,
  inlineStyleControls: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.arrayOf(function(
      propValue,
      key,
      componentName,
      location,
      propFullName
    ) {
      return propValue.some(p => {
        if (availableInlineStyles.indexOf(p) === -1) {
          return new Error(
            `Invalid prop ${propFullName} supplied to ${componentName}. Should be one of ${availableInlineStyles.join(
              " | "
            )}`
          )
        } else {
          return true
        }
      })
    })
  ]),
  listStyleControls: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.arrayOf(function(
      propValue,
      key,
      componentName,
      location,
      propFullName
    ) {
      return propValue.some(p => {
        if (availableListTypes.indexOf(p) === -1) {
          return new Error(
            `Invalid prop ${propFullName} supplied to ${componentName}. Should be one of ${availableListTypes.join(
              " | "
            )}`
          )
        } else {
          return true
        }
      })
    })
  ]),
  alignmentControls: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.arrayOf(function(
      propValue,
      key,
      componentName,
      location,
      propFullName
    ) {
      return propValue.some(p => {
        if (availableAlignmentStyles.indexOf(p) === -1) {
          return new Error(
            `Invalid prop ${propFullName} supplied to ${componentName}. Should be one of ${availableAlignmentStyles.join(
              " | "
            )}`
          )
        } else {
          return true
        }
      })
    })
  ])
}

EditorComponent.defaultProps = {
  blockStyleControls: true,
  dividerControl: true,
  inlineStyleControls: true,
  listControls: true,
  alignmentControls: true
}

export default EditorComponent
