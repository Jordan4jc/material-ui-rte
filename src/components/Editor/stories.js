import React from "react"
import { createMuiTheme } from "@material-ui/core/styles"
import { ThemeProvider } from "@material-ui/styles"
// Import the storybook libraries
import { storiesOf } from "@storybook/react"
import { action } from "@storybook/addon-actions"
// Import our component from this folder
import Editor from "./index"

const theme = createMuiTheme()

storiesOf("Editor").add("Basic", () => (
  <ThemeProvider theme={theme}>
    <Editor onChange={action("editor content changed")} />
  </ThemeProvider>
))
