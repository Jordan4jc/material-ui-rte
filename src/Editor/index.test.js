import React from "react"
import { shallow } from "enzyme"
import { unwrap } from "@material-ui/core/test-utils";
import Editor from "./index"

const EditorUnwrapped = unwrap(Editor);

describe("Editor", () => {
  it("Should render alignment control", () => {
    const component = shallow(<EditorUnwrapped classes={{}} />);
  
    expect(component).toMatchSnapshot();
  });
});