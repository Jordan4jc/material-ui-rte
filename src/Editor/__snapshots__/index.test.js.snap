// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor Should render alignment control 1`] = `
<WithStyles(ForwardRef(Paper))>
  <WithStyles(ForwardRef(Toolbar))>
    <BlockStyleControl
      onChange={[Function]}
    />
    <WithStyles(InlineStyleControls)
      editorState={
        EditorState {
          "_immutable": Immutable.Record {
            "allowUndo": true,
            "currentContent": Immutable.Record {
              "entityMap": Object {
                "__add": [Function],
                "__create": [Function],
                "__get": [Function],
                "__getLastCreatedEntityKey": [Function],
                "__mergeData": [Function],
                "__replaceData": [Function],
                "add": [Function],
                "create": [Function],
                "get": [Function],
                "getLastCreatedEntityKey": [Function],
                "mergeData": [Function],
                "replaceData": [Function],
              },
              "blockMap": Immutable.OrderedMap {
                "7fln5": Immutable.Record {
                  "key": "7fln5",
                  "type": "unstyled",
                  "text": "",
                  "characterList": Immutable.List [],
                  "depth": 0,
                  "data": Immutable.Map {},
                },
              },
              "selectionBefore": Immutable.Record {
                "anchorKey": "7fln5",
                "anchorOffset": 0,
                "focusKey": "7fln5",
                "focusOffset": 0,
                "isBackward": false,
                "hasFocus": false,
              },
              "selectionAfter": Immutable.Record {
                "anchorKey": "7fln5",
                "anchorOffset": 0,
                "focusKey": "7fln5",
                "focusOffset": 0,
                "isBackward": false,
                "hasFocus": false,
              },
            },
            "decorator": null,
            "directionMap": Immutable.OrderedMap {
              "7fln5": "LTR",
            },
            "forceSelection": false,
            "inCompositionMode": false,
            "inlineStyleOverride": null,
            "lastChangeType": null,
            "nativelyRenderedContent": null,
            "redoStack": Immutable.Stack [],
            "selection": Immutable.Record {
              "anchorKey": "7fln5",
              "anchorOffset": 0,
              "focusKey": "7fln5",
              "focusOffset": 0,
              "isBackward": false,
              "hasFocus": false,
            },
            "treeMap": Immutable.OrderedMap {
              "7fln5": Immutable.List [
                Immutable.Record {
                  "start": 0,
                  "end": 0,
                  "decoratorKey": null,
                  "leaves": Immutable.List [
                    Immutable.Record {
                      "start": 0,
                      "end": 0,
                    },
                  ],
                },
              ],
            },
            "undoStack": Immutable.Stack [],
          },
        }
      }
      onChange={[Function]}
    />
    <WithStyles(ListControls)
      editorState={
        EditorState {
          "_immutable": Immutable.Record {
            "allowUndo": true,
            "currentContent": Immutable.Record {
              "entityMap": Object {
                "__add": [Function],
                "__create": [Function],
                "__get": [Function],
                "__getLastCreatedEntityKey": [Function],
                "__mergeData": [Function],
                "__replaceData": [Function],
                "add": [Function],
                "create": [Function],
                "get": [Function],
                "getLastCreatedEntityKey": [Function],
                "mergeData": [Function],
                "replaceData": [Function],
              },
              "blockMap": Immutable.OrderedMap {
                "7fln5": Immutable.Record {
                  "key": "7fln5",
                  "type": "unstyled",
                  "text": "",
                  "characterList": Immutable.List [],
                  "depth": 0,
                  "data": Immutable.Map {},
                },
              },
              "selectionBefore": Immutable.Record {
                "anchorKey": "7fln5",
                "anchorOffset": 0,
                "focusKey": "7fln5",
                "focusOffset": 0,
                "isBackward": false,
                "hasFocus": false,
              },
              "selectionAfter": Immutable.Record {
                "anchorKey": "7fln5",
                "anchorOffset": 0,
                "focusKey": "7fln5",
                "focusOffset": 0,
                "isBackward": false,
                "hasFocus": false,
              },
            },
            "decorator": null,
            "directionMap": Immutable.OrderedMap {
              "7fln5": "LTR",
            },
            "forceSelection": false,
            "inCompositionMode": false,
            "inlineStyleOverride": null,
            "lastChangeType": null,
            "nativelyRenderedContent": null,
            "redoStack": Immutable.Stack [],
            "selection": Immutable.Record {
              "anchorKey": "7fln5",
              "anchorOffset": 0,
              "focusKey": "7fln5",
              "focusOffset": 0,
              "isBackward": false,
              "hasFocus": false,
            },
            "treeMap": Immutable.OrderedMap {
              "7fln5": Immutable.List [
                Immutable.Record {
                  "start": 0,
                  "end": 0,
                  "decoratorKey": null,
                  "leaves": Immutable.List [
                    Immutable.Record {
                      "start": 0,
                      "end": 0,
                    },
                  ],
                },
              ],
            },
            "undoStack": Immutable.Stack [],
          },
        }
      }
      onChange={[Function]}
    />
    <WithStyles(InlineAlignmentControls)
      editorState={
        EditorState {
          "_immutable": Immutable.Record {
            "allowUndo": true,
            "currentContent": Immutable.Record {
              "entityMap": Object {
                "__add": [Function],
                "__create": [Function],
                "__get": [Function],
                "__getLastCreatedEntityKey": [Function],
                "__mergeData": [Function],
                "__replaceData": [Function],
                "add": [Function],
                "create": [Function],
                "get": [Function],
                "getLastCreatedEntityKey": [Function],
                "mergeData": [Function],
                "replaceData": [Function],
              },
              "blockMap": Immutable.OrderedMap {
                "7fln5": Immutable.Record {
                  "key": "7fln5",
                  "type": "unstyled",
                  "text": "",
                  "characterList": Immutable.List [],
                  "depth": 0,
                  "data": Immutable.Map {},
                },
              },
              "selectionBefore": Immutable.Record {
                "anchorKey": "7fln5",
                "anchorOffset": 0,
                "focusKey": "7fln5",
                "focusOffset": 0,
                "isBackward": false,
                "hasFocus": false,
              },
              "selectionAfter": Immutable.Record {
                "anchorKey": "7fln5",
                "anchorOffset": 0,
                "focusKey": "7fln5",
                "focusOffset": 0,
                "isBackward": false,
                "hasFocus": false,
              },
            },
            "decorator": null,
            "directionMap": Immutable.OrderedMap {
              "7fln5": "LTR",
            },
            "forceSelection": false,
            "inCompositionMode": false,
            "inlineStyleOverride": null,
            "lastChangeType": null,
            "nativelyRenderedContent": null,
            "redoStack": Immutable.Stack [],
            "selection": Immutable.Record {
              "anchorKey": "7fln5",
              "anchorOffset": 0,
              "focusKey": "7fln5",
              "focusOffset": 0,
              "isBackward": false,
              "hasFocus": false,
            },
            "treeMap": Immutable.OrderedMap {
              "7fln5": Immutable.List [
                Immutable.Record {
                  "start": 0,
                  "end": 0,
                  "decoratorKey": null,
                  "leaves": Immutable.List [
                    Immutable.Record {
                      "start": 0,
                      "end": 0,
                    },
                  ],
                },
              ],
            },
            "undoStack": Immutable.Stack [],
          },
        }
      }
      onChange={[Function]}
    />
  </WithStyles(ForwardRef(Toolbar))>
  <div
    onClick={[Function]}
  >
    <DraftEditor
      blockRenderMap={
        Immutable.Map {
          "ordered-list-item": Object {
            "element": "li",
            "wrapper": <ol
              className="public-DraftStyleDefault-ol"
            />,
          },
          "header-six": Object {
            "element": "h6",
          },
          "header-four": Object {
            "element": "h4",
          },
          "header-one": Object {
            "element": "h1",
          },
          "unordered-list-item": Object {
            "element": "li",
            "wrapper": <ul
              className="public-DraftStyleDefault-ul"
            />,
          },
          "atomic": Object {
            "element": "figure",
          },
          "unstyled": Object {
            "aliasedElements": Array [
              "p",
            ],
            "element": "div",
          },
          "header-two": Object {
            "element": "h2",
          },
          "code-block": Object {
            "element": "pre",
            "wrapper": <pre
              className="public-DraftStyleDefault-pre"
            />,
          },
          "blockquote": Object {
            "element": "blockquote",
          },
          "header-five": Object {
            "element": "h5",
          },
          "header-three": Object {
            "element": "h3",
          },
        }
      }
      blockRendererFn={[Function]}
      blockStyleFn={[Function]}
      customStyleMap={
        Object {
          "ALIGN-CENTER": Object {
            "display": "block",
            "textAlign": "center",
          },
          "ALIGN-LEFT": Object {
            "display": "block",
            "textAlign": "left",
          },
          "ALIGN-RIGHT": Object {
            "display": "block",
            "textAlign": "right",
          },
        }
      }
      editorState={
        EditorState {
          "_immutable": Immutable.Record {
            "allowUndo": true,
            "currentContent": Immutable.Record {
              "entityMap": Object {
                "__add": [Function],
                "__create": [Function],
                "__get": [Function],
                "__getLastCreatedEntityKey": [Function],
                "__mergeData": [Function],
                "__replaceData": [Function],
                "add": [Function],
                "create": [Function],
                "get": [Function],
                "getLastCreatedEntityKey": [Function],
                "mergeData": [Function],
                "replaceData": [Function],
              },
              "blockMap": Immutable.OrderedMap {
                "7fln5": Immutable.Record {
                  "key": "7fln5",
                  "type": "unstyled",
                  "text": "",
                  "characterList": Immutable.List [],
                  "depth": 0,
                  "data": Immutable.Map {},
                },
              },
              "selectionBefore": Immutable.Record {
                "anchorKey": "7fln5",
                "anchorOffset": 0,
                "focusKey": "7fln5",
                "focusOffset": 0,
                "isBackward": false,
                "hasFocus": false,
              },
              "selectionAfter": Immutable.Record {
                "anchorKey": "7fln5",
                "anchorOffset": 0,
                "focusKey": "7fln5",
                "focusOffset": 0,
                "isBackward": false,
                "hasFocus": false,
              },
            },
            "decorator": null,
            "directionMap": Immutable.OrderedMap {
              "7fln5": "LTR",
            },
            "forceSelection": false,
            "inCompositionMode": false,
            "inlineStyleOverride": null,
            "lastChangeType": null,
            "nativelyRenderedContent": null,
            "redoStack": Immutable.Stack [],
            "selection": Immutable.Record {
              "anchorKey": "7fln5",
              "anchorOffset": 0,
              "focusKey": "7fln5",
              "focusOffset": 0,
              "isBackward": false,
              "hasFocus": false,
            },
            "treeMap": Immutable.OrderedMap {
              "7fln5": Immutable.List [
                Immutable.Record {
                  "start": 0,
                  "end": 0,
                  "decoratorKey": null,
                  "leaves": Immutable.List [
                    Immutable.Record {
                      "start": 0,
                      "end": 0,
                    },
                  ],
                },
              ],
            },
            "undoStack": Immutable.Stack [],
          },
        }
      }
      keyBindingFn={[Function]}
      onChange={[Function]}
      readOnly={false}
      spellCheck={false}
      stripPastedStyles={false}
    />
  </div>
</WithStyles(ForwardRef(Paper))>
`;
